### ISSUE TITLE
Add Wavpack support

### ISSUE THREAD

**Issue by erikd** [libsndfile/libsndfile#15]
Obvious enough.

Unfortunately it is unlikely that I (Erik) will find sufficient time or motivation to do this, but I would be happy to accept patches.

**Comment by phoenixx2** [comment]
Any progress?  Will need to implement immediately otherwise.  Still open for patches?

**Comment by daurnimator** [comment]
> Any progress?

Sorry, I didn't get enough time to work on this :(


> Still open for patches?

Please go ahead and write it yourself if you have the time.

**Comment by phoenixx2** [comment]
Are there any build instructions for Windows or someone have a solution going?  I'm not familiar with CMake but as far as I can see there's only Linux instructions referencing autogen.sh and running cmake won't work until that script has been executed.  Excuse my ignorance.

**Comment by phoenixx2** [comment]
Never mind - I can't use this project on iOS due to LGPL.  Good luck.

**Comment by erikd** [comment]
@phoenixx2 Just for the record, libsndfile *can* be used on iOS, but complying with the LGPL means that any code that statically links to libsndfile must be releaed under an LGPL compaible license. 

An example of such a project is csound-for-ios : http://www.csounds.com/shop/csound-for-ios/

**Comment by martinwguy** [comment]
Is using libwavpack an acceptable solution to this?

I see in configure.ac:
"One or more of the external libraries (ie libflac, libogg and libvorbis) is either missing
"Unfortunately, for ease of maintenance, the external libs are an all or nothing affair".

so if I tried to include libwavpack support via libwavpack within the current configure/build framework, that mean that, on systems without libwavpack development packages installed, that would also make flac, ogg and vorbis support unusable?

How does insisting on all-or-nothing external libraries improve ease of maintenance, and is fixing *that* before adding wavpack support via libwavpack acceptable? That would more easily enable other formats with external library support (like libopus, issue #69).

The alternative, of hacking down and including other libraries' sources in libsndfile, then having to update the modified versions every time they make bug fixes upstream, seems worse still for maintenance and in practice, no upstream bug fixes and improvements would probably ever get applied to their libsndfile forks, however critical.

**Comment by erikd** [comment]
> Is using libwavpack an acceptable solution to this?

That's probably the only acceptable solution.

Ogg, Vorbis and FLAC are supported as a block because they are so intertwined. Ogg without Vorbis or Vorbis without Ogg support is pretty close to meaningless. I lumped FLAC in as well because libsndfile support FLAC data in an Ogg container.

Since there is no interdependance for wavpack, I'm perfectly happy with making wavpack completely independent (and fixing the comments in `configure.ac`.

> How does insisting on all-or-nothing external libraries improve ease of maintenance?

It reduces tha mintenance burden because only two configurations need to be tested; FLAC/Ogg/Vorbis all enabled and FLAC/Ogg/Vorbis all disabled. Just splitting that into three separate configurations results in the following configurations (all of which need to be tested):

* None.
* FLAC alone (making sure Ogg/Vorbis and Ogg/FLAC are disabled).
* FLAC with Ogg (making sure Ogg/Vorbis is disabled).
* Ogg/Vorbis alone (making sure FLAC and Ogg/FLAC are disabled).
* FLAC and Ogg/Vorbis.

That is now 5 combinations that need to be tested. 

On top of that, getting this extra configurability can really only be done with `#ifdef` hackery which in my experience is both fragile and easy to get wrong. In addition, the changes required to support this affect the code in a number of places other than the just the Ogg/Vorbis/FLAC support code. To make sure this is and remains correct, more test code is required.

> Is it OK to separate FLAC/Ogg/Vorbis support in configure before adding wavpack support via libwavpack?

I'd be happy to rename rename `HAVE_EXTERNAL_LIBS` to `HAVE_EXTERBAL_XIPH_LIBS` and make wavpack support completely separate. In fact, I think I will do that anyway.

> That would also make it possible to add other formats with external library support (like libopus issue #69 and to finish the libspeex support

Because both Opus and Speex *require* Ogg as the container,  think they should be lumped in with FLAC/Ogg/Vorbis.

**Comment by martinwguy** [comment]
It turns out that Ogg can occur in files without Vorbis (e.g. Ogg-Opus), and Vorbis without Ogg (in Matroska or WebM) but that's totally irrelevant to compiling libsndfile unless people are making an embedded solution that is short of program space (there is one distro whose build script modifies libsndfile to get ogg/vorbis without flac, presumably for this reason).

I think FLAC is always in an Ogg container too, so I guess libflac does that internally, without using libogg. I dunno. Maybe libspeex does the same thing, as it decodes speex files without linking to libogg.

Anyway, in the meantime I've tried separating into three flags, with testsuite changes, and all 8 combinations now configure and make-check fine. It turns out that there is no interdependence in libsndfile between the presence of the three libraries, and the presence of each just enables its own code fragments, by changing #if HAVE_EXTERNAL_LIBS to HAVE_LIBFLAC or whatever, only very rarely needing (HAVE_LIBOGG && HAVE_LIBVORBIS) in a couple of places such as the list of simple formats.

I've uploaded my hackery and made a pull request, but only because I'm signing off now to go out.
If it's sufficient, take it and I'll try lib* tomorrow based on this. If not, let me know what you prefer and I'll rework it instead.  Thanks!

**Comment by erikd** [comment]
FLAC has its own container format as well as supporting Ogg as a container format.

> Anyway, in the meantime I've tried separating into three flags

And I just push commit 95828ca2e9.

**Comment by martinwguy** [comment]
:D

**Comment by erikd** [comment]
The configuration of Ogg/Vorbis and FLAC is actually completely orthogonal to Wavpack support.

**Comment by tuxzz** [comment]
> The configuration of Ogg/Vorbis and FLAC is actually completely orthogonal to Wavpack support.

Not really. FLAC doesn't support float32 format as input data but Wavpack does. Float32 is useful when we storage raw tracks for future audio processing, or scientific audio data for SNR etc. comutation.

**Comment by evpobr** [comment]
No doubt WavPack is in wishlist. We still have no developer to implement it.

**Comment by tuxzz** [comment]
I'm not familiar with codebase of libsndfile. But I'm willing to take this task because I need to use it recently :P

**Comment by evpobr** [comment]
Good. Create pull request. You can start from [How to add new file format](https://github.com/libsndfile/libsndfile/blob/master/docs/new_file_type_howto.md) and [`src/new.c`](https://github.com/libsndfile/libsndfile/blob/master/src/new.c).
